crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -342,6 +342,21 @@ impl SyncedKeyQueryManager<'_> {
         self.cache.tracked_users.read().iter().cloned().collect()
     }
 
+    /// See the docs for
+    /// [`crate::OlmMachine::is_user_tracked_and_up_to_date()`].
+    pub async fn is_user_tracked_and_up_to_date(&self, user_id: &UserId) -> bool {
+        let is_tracked = self.cache.tracked_users.read().contains(user_id);
+
+        if is_tracked {
+            // If we're tracked, we want to know if we're queued up for a `/keys/query`.
+            //
+            // Being in this map means that we're not up-to-date.
+            !self.manager.users_for_key_query.lock().await.contains_user(user_id)
+        } else {
+            false
+        }
+    }
```

Comment:
I'm sorry about this, because I suspect it's going to make this scope-creep even more, and maybe just doing a `/keys/query` every time was the lesser evil, but...

I don't think this is sufficient. If I *previously* shared a room with Bob, but have now left, then Bob will still be present in `tracked_users` (I think?), but the device list may well be outdated.

Reply:
yeah, sorry, trying to optimise away the `/keys/query` was a terrible idea.

Turns out cache invalidation is hard. Who knew.